### PR TITLE
swanhub: new `conda` builder option

### PR DESCRIPTION
**Pull request by RodrigoSobral2000**
Swanhub can now also forward `conda` builder option through URL param to `SwanCustomEnvironments` extension

### CODE REVIEW

File: SwanHub/swanhub/spawn_handler.py [hunk 1]
```diff
@@ -173,7 +173,8 @@ def _finish_spawn(f):
                 "repo_type": options.get(configs.repo_type),
                 "file": options.get(configs.file, ''),
             }
-            if options.get(configs.builder) == configs.accpy_special_type:
+            special_builders = [configs.accpy_special_type, configs.conda_special_type]
```
Comment (etejedor) [inline SwanHub/swanhub/spawn_handler.py:176]: What does "special" mean here? What is the particular treatment you do for this two? This would need a comment.

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: Means that a builder different from the default one (`python-venv`) will be sent

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: I intend to change this logic along with the extension in the future, to make it more modular and flexible.
But for now I think is better to keep it this way

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: Since we are adding conda, let's just do it the way we think it's right from the beginning, no? Why reviewing something that is not final, to do another review round afterwards?

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: Also, why is the assignment below only done for conda and acc-py?

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: I think is better to change it in a separate commit because the changes I'm planning to make will probably have further implications. Turning the PR more generic and not properly conda focused.

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: Currently, unless a `conda` or `acc-py` is explicitly specified, the `makenv` script will build the environment with `python-venv`.

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: Yes, but why `query_params[options.get(configs.builder)] = options.get(configs.builder_version)` is only done for conda and accpy?

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: Yes, that's it.
`query_params` will be `{..., 'accpy': '2023.06', ...}` and the `makenv` script will have `--builder=builder_version`, this is `--accpy=2023.06` or `--conda=mamba`

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: But, as mentioned, I think would be better to change the script in order to do `--builder=accpy --builder_version=2023.06` for instance

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: > But, as mentioned, I think would be better to change the script in order to do --builder=accpy --builder_version=2023.06 for instance

I agree with this.

> this is --accpy=2023.06 or --conda=mamba

So the `conda` builder has a version called `mamba`? I do not think it is necessary for the conda builder to have a version?

Bottom line, query_params would container `builder_version` only in the case of accpy.

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: I know for now either python or conda use only one "version" (venv and mamba). But for avoiding more if and elses, and even for keeping the consist format of `builder-version` on form options, I think would be better to keep this `builder_version` (and make it mandatory on the extension). This way we are also preparing ourselves for future cases like the ones we discussed about the experiments.

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: You say that the venv builder also has a version called venv? But then why the assignment here is not done for venvs? I'm a bit lost.

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: The `builder-versions` we got so far are:
- `python3-venv`
- `conda-mamba`
- `accpy-2023.06`
- `accpy-2021.12`
- `accpy-2020.11`

Im trying to follow a pattern of `<binary>-<distribution/version>`

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: conda-mamba is ok, for venv I would revert it (venv-python3).

But you still haven't answered my question :) Why is the assignment only done for `special_builders` that are just conda and accpy, when you are saying that all the builders have version?

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: Because the `python3-venv` was the default `makenv` behaviour, which means you didn't even need to specified it

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:176]: It was a dummy approach, but that has changed now.

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:176]: Ok. Then what I would do is not have this special variable. I would always do the assignment, with whatever comes from the form -- for accpy it will be a certain version, for venv it will always be python3, for conda it will always be mamba.

File: SwanSpawner/swanspawner/templates/options_form_template.html [hunk 2]
```diff
@@ -325,7 +325,7 @@
             if (urlParams.get('file') !== null  && urlParams.get('file') !== '') {
                 toggle_visibility('file_config');
             }
-        } else {
+        } else if (software_source) {
```
Comment (etejedor) [inline SwanSpawner/swanspawner/templates/options_form_template.html:328]: You put this back because it is necessary?

Comment (RodrigoSobral2000) [reply SwanSpawner/swanspawner/templates/options_form_template.html:328]: Yes it is. If no `software_source` is specified, means the user is normally accessing the form, with no query arguments. So, there's no reason to show the error message. We only want to show it if the `software_source` is specified but incorrect

Comment (etejedor) [reply SwanSpawner/swanspawner/templates/options_form_template.html:328]: Can't it happen that the user did specify a software source via URL args, but it was empty?

Comment (RodrigoSobral2000) [reply SwanSpawner/swanspawner/templates/options_form_template.html:328]: It can, but in that case (`?software_source=`), it will be ignored

Comment (etejedor) [reply SwanSpawner/swanspawner/templates/options_form_template.html:328]: It will be ignored and you won't notify to the user that it was wrong?

Comment (RodrigoSobral2000) [reply SwanSpawner/swanspawner/templates/options_form_template.html]: This was address in a way that `?software_source=` triggers an error too, letting the user know the argument must not be empty

File: SwanSpawner/swanspawner/templates/options_form_template.html [hunk 3]
```diff
@@ -325,7 +325,7 @@
             if (urlParams.get('file') !== null  && urlParams.get('file') !== '') {
                 toggle_visibility('file_config');
             }
-        } else {
+        } else if (software_source !== null) {
```
Comment (etejedor) [inline SwanSpawner/swanspawner/templates/options_form_template.html:328]: Is this related to the conda builder options? If not, it needs to be in a separate commit, which explains why this change is needed.

File: SwanHub/swanhub/spawn_handler.py [hunk 4]
```diff
@@ -171,10 +171,10 @@ def _finish_spawn(f):
             query_params = {
                 "repo": options.get(configs.repository),
                 "repo_type": options.get(configs.repo_type),
+                "builder": options.get(configs.builder),
+                "builder_version": options.get(configs.builder_version),
```
Comment (etejedor) [inline SwanHub/swanhub/spawn_handler.py:175]: iuuc `builder_version` will be set here as an empty string for `venv` and `mamba`?

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:175]: Yes, we are always sending it to the api. But it can become even more optional.

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py:175]: Nevermind, I forgot that I have this thing of making `builder_version` optional done in a separated branch cause I was thinking about implementing it in a separated PR too. Could I include these changes in this PR (different commits tho)

Comment (etejedor) [reply SwanHub/swanhub/spawn_handler.py:175]: Yes please.

Comment (RodrigoSobral2000) [reply SwanHub/swanhub/spawn_handler.py]: Done, I think all changes that were needed are commited

File: SwanHub/swanhub/spawn_handler.py [hunk 5]
```diff
@@ -171,10 +171,11 @@ def _finish_spawn(f):
             query_params = {
                 "repo": options.get(configs.repository),
                 "repo_type": options.get(configs.repo_type),
+                "builder": options.get(configs.builder),
                 "file": options.get(configs.file, ''),
             }
-            if options.get(configs.builder) == configs.accpy_special_type:
-                query_params[options.get(configs.builder)] = options.get(configs.builder_version)
+            if options.get(configs.builder_version):
```
Comment (etejedor) [inline SwanHub/swanhub/spawn_handler.py:177]: Can you add a comment here:
```
# If the builder has a version, pass it as an argument of the query
```

File: SwanSpawner/swanspawner/swanspawner.py [hunk 6]
```diff
@@ -113,10 +115,13 @@ def options_from_form(self, formdata):
                 if not options[self.repository]:
                     raise ValueError('Cannot create custom software environment: no repository specified')
 
-                # Builders are specified in builder-builderversion format
-                builder, builder_version = formdata[self.builder][0].lower().split('-')
-                options[self.builder]         = builder
-                options[self.builder_version] = builder_version
+                # Builders are specified in builder:builderversion format (e.g. swan:1.0)
```
Comment (etejedor) [inline SwanSpawner/swanspawner/swanspawner.py:118]: But now you have builder and builder_options, and the latter is only set for acc-py, right? Why matching a pattern on the builder, which now can only be `venv`, `mamba` or `acc-py`?

Comment (etejedor) [reply SwanSpawner/swanspawner/swanspawner.py:118]: Ok, now I understand.

**Review by etejedor — APPROVED** [review]
